### PR TITLE
Implement ES2021 WeakRef

### LINKED ISSUE #943: Support EcmaScript 2021 WeakRef & FinalizationRegistry
> WeakRef, for referring to a target object without preserving it from garbage collection
> FinalizationRegistry, to manage registration and unregistration of cleanup operations performed when target objects are garbage collected

See [MDN WeakRef](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef)

### PR DESCRIPTION
**Pull request by anivar**
## Summary
Implements ES2021 WeakRef for Rhino JavaScript engine.

## Changes
- **NativeWeakRef.java**: Core WeakRef implementation using Java's WeakReference
- **ScriptRuntime.java**: Register WeakRef constructor for ES6+ language versions  
- **Messages.properties**: Add error messages for invalid WeakRef constructor arguments
- **WeakRefTest.java**: Comprehensive test suite covering all edge cases

## Implementation Details
- Uses Java's `java.lang.ref.WeakReference` internally for proper weak reference semantics
- Follows Rhino's established LambdaConstructor pattern (consistent with NativePromise)
- Supports WeakRef constructor validation and `deref()` method per ES2021 spec
- Throws TypeError for invalid arguments (primitives, null, undefined, no arguments)
- Provides correct `[object WeakRef]` toString tag via Symbol.toStringTag

## Testing
- All custom WeakRef tests pass in both interpreted and compiled modes
- Maintains backward compatibility with existing code

## Use Cases
WeakRef allows holding weak references to objects without preventing garbage collection, useful for:
- Memory-sensitive caches and mappings
- Observer patterns without memory leaks
- Cleanup callbacks and resource management

Closes #943

### CODE REVIEW

**Comment by gbrail** [comment]
Thank you for doing this!

Moving to a new version of test262 is a big deal and touches a lot of files -- but it's worth it so we can stay on top of the project?Could you try doing that in a different PR? Then your PR can just focus on the new feature.

**Comment by anivar** [comment]
Thanks for the feedback\! I've separated the test262 update into PR #2055. This PR now focuses only on the WeakRef implementation.

**Comment by rbri** [comment]
@anivar 
### test262.properties!!!!

File: rhino/src/main/resources/org/mozilla/javascript/resources/Messages.properties [hunk 1]
```diff
@@ -1021,6 +1021,22 @@ msg.promise.all.toobig =\
...
   Too many inputs to Promise.any
 
+# WeakRef and FinalizationRegistry
+msg.weakref.no.target =\
+  WeakRef constructor requires an object argument
+
+msg.weakref.target.not.object =\
+  WeakRef target must be an object
+
+msg.finalization.registry.no.callback =\
```
Comment (rbri) [inline rhino/src/main/resources/org/mozilla/javascript/resources/Messages.properties:1031]: @anivar please handle the finalizationregistry stuff in a separate PR

File: rhino/src/main/java/org/mozilla/javascript/ScriptRuntime.java [hunk 2]
```diff
@@ -257,6 +257,11 @@ public static ScriptableObject initSafeStandardObjects(
             new LazilyLoadedCtor(scope, "Reflect", sealed, true, NativeReflect::init);
         }
 
+        // ES2021 features
```
Comment (rbri) [inline rhino/src/main/java/org/mozilla/javascript/ScriptRuntime.java:260]: maybe this can be part of the if above

File: tests/src/test/java/org/mozilla/javascript/tests/es2025/WeakRefTest.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.javascript.tests.es2025;
+
+import org.junit.Test;
+import org.mozilla.javascript.testutils.Utils;
+
+public class WeakRefTest {
+
+    // ========== UTILITY METHODS ==========
```
Comment (rbri) [inline tests/src/test/java/org/mozilla/javascript/tests/es2025/WeakRefTest.java:12]: as far as i remember these kind of grouping are not used in the codebase before

File: tests/src/test/java/org/mozilla/javascript/tests/es2025/WeakRefTest.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+    private void assertES6(String script, Object expected) {
+        Utils.assertWithAllModes_ES6(expected, script);
```
Comment (rbri) [inline tests/src/test/java/org/mozilla/javascript/tests/es2025/WeakRefTest.java:15]: does this method really make sense?

File: tests/src/test/java/org/mozilla/javascript/tests/es2025/WeakRefTest.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    private void assertES18(String script, Object expected) {
+        Utils.assertWithAllModes_1_8(expected, script);
+    }
+
+    private void assertTypeError(String script) {
```
Comment (rbri) [inline tests/src/test/java/org/mozilla/javascript/tests/es2025/WeakRefTest.java:22]: Maybe something we can move to the Utils? Or we have to align a bit about error testing....
